### ISSUE TITLE
Couldn't receive callback from postponed transaction.

### ISSUE THREAD

**Issue by joyeon** [chirag04/react-native-in-app-utils#20]
Hello there,

Thanks for this useful plugin!

I found a react-native app couldn't receive the callback properly when itunes store postpones a transaction because of some reason like user verification, but It asks ths user to purchase an item again after verification. Because when itunes store postpones a transaction, It drops `SKPaymentTransactionStateFailed` first and the native module remove the callback at that time.

I tried to remove callback only when the user cancels the transaction. But I'm not sure It is proper way.

Thanks,

**Comment by chirag04** [comment]
@joyeon Does apple post two events one `SKPaymentTransactionStateFailed` and something else later after verification?

**Comment by joyeon** [comment]
Yes apple ask to let user purchase again after verification. If the user accept it, apple post an event `SKPaymentTransactionStatePurchased`.

**Comment by chirag04** [comment]
Interesting. Not sure what's the correct way to handle this. We should not leak callbacks when we get `SKPaymentTransactionStateFailed`.

Maybe it's time to move away from `callback` based pattern to an event based one. @joyeon you want to send a PR?

**Comment by joyeon** [comment]
Will try to send PR but It could take time. If you want to patch it immediately please do it without hesitation.

**Comment by mcrowe** [comment]
I have been experiencing this same issue. I believe this is because this library is focused on a callback pattern, which only listens for transaction changes during the lifetime of that request. Best practices are to listen for transactions *when the app launches* so that postponed requests like this are seen. See [Apple's description](https://developer.apple.com/library/content/technotes/tn2387/_index.html#//apple_ref/doc/uid/DTS40014795-CH1-BEST_PRACTICES-ADD_A_TRANSACTION_QUEUE_OBSERVER_AT_APPLICATION_LAUNCH) for more details.

This happens quite rarely, and I'm not sure how to recreate it deliberately. We get about 750 new subscriptions per month, and this seems to happen to about 1 user per month. Still, it leads to a pretty frustrating experience for that 1 user.

As @chirag04 suggests, I think the solution lies in moving towards an event-based architecture. I will be working on a solution imminently. I may provide a PR, or create a new library, depending on the extent of the changes needed.

**Comment by mcrowe** [comment]
This is related to #31

**Comment by lefoy** [comment]
Any news on this? I'm having the same issue
